drivers: counter: add st,stm32-counter to stm32h7

XenuIsWatching opened the pull request:
The STM32H7 was missing definitions in it's devicetree include for the stm32-counter

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

XenuIsWatching committed:
drivers: counter: add st,stm32-counter to stm32h7

The STM32H7 was missing definitions in it's devicetree include
for the stm32-counter

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

erwango reviewed (changes requested):
Thanks for this change. This looks correct and I guess you tested it.
However, it'd be needed to enable on some upstream H7 platform so that:
- anyone interested can verify it quickly.
- it could be tested in CI (using `tests/drivers/counter/counter_basic_api`).

Can you arrange this ?

FRASTM commented:
looks similar to https://github.com/zephyrproject-rtos/zephyr/pull/52426
In the end LL_TIM_OC_GetCompareCHx(**const** TIM_TypeDef  is defined for stm32f7, stm32g0, stm32l4, stm32wb
Should this be considered in the drivers/counter/counter_ll_stm32_timer.c globally ?

erwango commented:
> Should this be considered in the drivers/counter/counter_ll_stm32_timer.c globally ?

I prefer this is enabled case by case with matching device tree configuration and device tree configuration.

XenuIsWatching commented:
> it'd be needed to enable on some upstream H7 platform

Sorry, I'm not quite familiar with this. Could you point me to an example?

erwango commented:
> > it'd be needed to enable on some upstream H7 platform
> 
> Sorry, I'm not quite familiar with this. Could you point me to an example?

Something like: https://github.com/zephyrproject-rtos/zephyr/commit/8e79af0ecaa9f7adfe68801b36965d98f65ac8df

XenuIsWatching committed:
tests: drivers: counter: basic_api: add overlay for stm32h747i_disco_m7

Add overlay for stm32h747i_disco_m7 to enable counter_basic_api
tests on this board. The prescalers of the timers will result
in a counter frequency of 1 MHz.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

erwango reviewed (changes requested):
Test won't work as board name is not correct:
stm32h747i_disco.overlay -> stm32h747i_disco_m7.overlay

Besides, you need to add `counter` as supported feature in board's yaml file.

I've done these change and I'm getting:
```
devicetree error: ~/zephyrproject/zephyr/tests/drivers/counter/counter_basic_api/boards/stm32h747i_disco_m7.overlay:43 (column 1): parse error: undefined node label 'timers9'
```

GeorgeCGV commented on tests/drivers/counter/counter_basic_api/boards/stm32h747i_disco_m7.overlay:
```diff
@@ -0,0 +1,87 @@
+		status = "okay";
+	};
+};
+
```
Shouldn't `timers8` be enabled?

XenuIsWatching replied:
`timers1` and `timers8` are the Advanced-control timers and work different that driver already written

erwango reviewed (changes requested):
Looks ok but test is failed on my board:
```
START - test_single_shot_alarm_top
Testing counter
Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:411: test_single_shot_alarm_instance: (1 not equal to cnt)
counter: Expecting alarm callback
FAIL - test_single_shot_alarm_top in 0.059 seconds
===================================================================
TESTSUITE counter_basic failed.
```

Would you mind investigating ?
(Note that the test is passed on L4 based board)

XenuIsWatching commented:
> Looks ok but test is failed on my board:
> 
> ```
> START - test_single_shot_alarm_top
> Testing counter
> Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/counter/counter_basic_api/src/test_counter.c:411: test_single_shot_alarm_instance: (1 not equal to cnt)
> counter: Expecting alarm callback
> FAIL - test_single_shot_alarm_top in 0.059 seconds
> ===================================================================
> TESTSUITE counter_basic failed.
> ```
> 
> Would you mind investigating ? (Note that the test is passed on L4 based board)

ok, I took a look at it, and I noticed an oddity in the test suite.  Apparently if the top value and alarm value equal each other, the tim won't count up to the exact value.... but it will count to that value-1. In the test there is a set to `alarm_cfg.ticks = ticks - 1;` [1] ... but then the very next line, it does a `alarm_cfg.ticks = ticks;` ... seems odd and unnecessary?? Having `alarm_cfg.ticks = ticks - 1;` when `set_top` is `true` passes the test case. I'd have to do some digging in the reference manual of the OC reg can equal the ARR...

[1] https://github.com/zephyrproject-rtos/zephyr/blob/a1b4896efe46772ac0f6c84976e1b17b9dd6dfae/tests/drivers/counter/counter_basic_api/src/test_counter.c#L402

[2] https://github.com/zephyrproject-rtos/zephyr/blob/a1b4896efe46772ac0f6c84976e1b17b9dd6dfae/tests/drivers/counter/counter_basic_api/src/test_counter.c#L405

There was also another oddity in the driver I noticed. I've got a project where I am using the ITRs of the timers where one timer can trigger another timer wtih the OC1 as the event. I'm using the STM32 LL libraries for setting this around the zephyr driver, but when ever I set an alarm for the 1st channel, I noticed that the ITR of the slave timer gets triggered immeditally with the function call, and then triggered again when CNT reaches the OC1 value. I did some digging and found that some "odd" code [3] within the set_cc is causing this issue

FRASTM commented:
Please @XenuIsWatching  : Fix conflict + consider @gautierg-st remarks

XenuIsWatching commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -36,7 +36,7 @@ static void(*const set_timer_compare[TIMER_MAX_CH])(TIM_TypeDef *,
 
 /** Channel to compare get function mapping. */
-#if !defined(CONFIG_SOC_SERIES_STM32F4X)
+#if !defined(CONFIG_SOC_SERIES_STM32F4X) && !defined(CONFIG_SOC_SERIES_STM32H7X)
```
Can you point me to where that update was, specifically the commit

gautierg-st replied:
It is this commit in the hal_stm32 repo:
https://github.com/zephyrproject-rtos/hal_stm32/commit/55648728d4ca9d17595d6adc2db447438c48bfef
It updates the prototypes of STM32H7 timer function to use const TIM_TypeDef, hence your modification should not be needed.
To use the latest HAL version, you should make a `west  update` to update your whole modules folder.

XenuIsWatching commented:
Conflict fixed... but I need more clarification on @gautierg-st remark

FRASTM commented:
> Conflict fixed... but I need more clarification on @gautierg-st remark

when running the $ west update , you will get the latest hal_stm32 where the sm32h7xx_ is already modified

FRASTM commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -37,7 +37,7 @@ static void(*const set_timer_compare[TIMER_MAX_CH])(TIM_TypeDef *,
 
 /** Channel to compare get function mapping. */
-#if !defined(CONFIG_SOC_SERIES_STM32F4X)
+#if !defined(CONFIG_SOC_SERIES_STM32F4X) && !defined(CONFIG_SOC_SERIES_STM32H7X)
```
```suggestion
#if !defined(CONFIG_SOC_SERIES_STM32F4X)
```

FRASTM commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -62,7 +62,7 @@ static void(*const disable_it[TIMER_MAX_CH])(TIM_TypeDef *) = {
 #ifdef CONFIG_ASSERT
 /** Channel to interrupt enable check function mapping. */
-#if !defined(CONFIG_SOC_SERIES_STM32F4X)
+#if !defined(CONFIG_SOC_SERIES_STM32F4X) && !defined(CONFIG_SOC_SERIES_STM32H7X)
```
```suggestion
#if !defined(CONFIG_SOC_SERIES_STM32F4X)
```

erwango commented on boards/arm/stm32h747i_disco/stm32h747i_disco_m7.yaml:
```diff
@@ -17,3 +17,4 @@ supported:
   - memc
   - usb_cdc
   - usb_device
+  - counter
```
If declaring counter is supported in the target, for coherency, you should at least enable a node in .dts.
If we start having divergence between .dts and .yaml it will make things quite hard to track.

XenuIsWatching replied:
I see this `disabled` in all of the stm32*.dtsi that support `st,stm32-counter` as well as seeing them in their board yaml, so I'm at least consistent 🤷 .... but I don't want to just enable it if there isn't a use for it as a generic timer.  I could just remove this line then...

erwango replied:
The effect is now that counter won't be tested in CI on this board due to missing yaml "counter" dependency.
I'm approving anyway because we need to progress, but this will have to be fixed somehow
